File: backend/services/agent_filesystem_service.py
"""
Agent filesystem service.

Handles agent file operations (create, archive) for worlds.
"""

import logging
import os
import shutil
from datetime import datetime
from typing import Optional

from services.world_service import WorldService

logger = logging.getLogger("AgentFilesystemService")


class AgentFilesystemService:
    """Agent filesystem management service."""

    @classmethod
    def create_agent(
        cls,
        world_name: str,
        agent_name: str,
        in_a_nutshell: str,
        characteristics: str,
        profile_pic_path: Optional[str] = None,
    ) -> None:
        """Create a new character agent in the world."""
        world_path = WorldService.get_world_path(world_name)
        agent_path = world_path / "agents" / agent_name

        agent_path.mkdir(exist_ok=True)

        with open(agent_path / "in_a_nutshell.md", "w", encoding="utf-8") as f:
            f.write(in_a_nutshell)

        with open(agent_path / "characteristics.md", "w", encoding="utf-8") as f:
            f.write(characteristics)

        if profile_pic_path and os.path.exists(profile_pic_path):
            ext = os.path.splitext(profile_pic_path)[1]
            shutil.copy(profile_pic_path, agent_path / f"profile{ext}")

        logger.info(f"Created agent '{agent_name}' in world '{world_name}'")

    @classmethod
    def archive_agent(cls, world_name: str, agent_name: str) -> bool:
        """
        Archive a character agent (move to archived folder).

        Used when a character dies, departs, or is manually removed.

        Args:
            world_name: Name of the world
            agent_name: Name of the agent to archive

        Returns:
            True if archived successfully, False if agent not found
        """
        world_path = WorldService.get_world_path(world_name)
        agent_path = world_path / "agents" / agent_name
        archived_path = world_path / "agents" / "_archived"

        if not agent_path.exists():
            logger.warning(f"Agent '{agent_name}' not found in world '{world_name}'")
            return False

        # Create archived directory if it doesn't exist
        archived_path.mkdir(exist_ok=True)

        # Move agent folder to archived (with timestamp to avoid collisions)
        timestamp = datetime.utcnow().strftime("%Y%m%d_%H%M%S")
        archived_agent_path = archived_path / f"{agent_name}_{timestamp}"

        shutil.move(str(agent_path), str(archived_agent_path))
        logger.info(f"Archived agent '{agent_name}' in world '{world_name}' to {archived_agent_path}")
        return True

    @classmethod
    def list_world_agents(cls, world_name: str) -> list[str]:
        """
        List all agent names in a world's agents folder.

        Excludes archived agents and internal folders (starting with _).

        Args:
            world_name: Name of the world

        Returns:
            List of agent names (folder names)
        """
        world_path = WorldService.get_world_path(world_name)
        agents_path = world_path / "agents"

        if not agents_path.exists():
            return []

        agent_names = []
        for item in agents_path.iterdir():
            # Skip non-directories and internal folders (e.g., _archived)
            if item.is_dir() and not item.name.startswith("_"):
                # Verify it's a valid agent folder (has at least in_a_nutshell.md)
                if (item / "in_a_nutshell.md").exists() or (item / "characteristics.md").exists():
                    agent_names.append(item.name)

        return agent_names

    @classmethod
    def get_agent_details(cls, world_name: str, agent_name: str) -> Optional[dict]:
        """
        Get agent details from filesystem.

        Args:
            world_name: Name of the world
            agent_name: Name of the agent folder

        Returns:
            Dict with agent details or None if not found:
            - name: Agent name (display name, underscores replaced with spaces)
            - in_a_nutshell: Brief description
            - profile_pic: Path to profile picture (if exists)
        """
        world_path = WorldService.get_world_path(world_name)
        agent_path = world_path / "agents" / agent_name

        if not agent_path.exists():
            return None

        # Read in_a_nutshell
        in_a_nutshell = ""
        nutshell_file = agent_path / "in_a_nutshell.md"
        if nutshell_file.exists():
            try:
                in_a_nutshell = nutshell_file.read_text(encoding="utf-8").strip()
            except Exception as e:
                logger.warning(f"Failed to read in_a_nutshell.md for {agent_name}: {e}")

        # Find profile picture
        profile_pic = None
        profile_names = ["profile", "avatar", "picture", "photo"]
        profile_exts = [".png", ".jpg", ".jpeg", ".gif", ".webp", ".svg"]
        for name in profile_names:
            for ext in profile_exts:
                pic_path = agent_path / f"{name}{ext}"
                if pic_path.exists():
                    profile_pic = str(pic_path)
                    break
            if profile_pic:
                break

        return {
            "name": agent_name,
            "folder_name": agent_name,
            "in_a_nutshell": in_a_nutshell,
            "profile_pic": profile_pic,
        }

    @classmethod
    def list_world_agents_with_details(cls, world_name: str) -> list[dict]:
        """
        List all agents in a world with their details.

        Args:
            world_name: Name of the world

        Returns:
            List of agent detail dicts
        """
        agent_names = cls.list_world_agents(world_name)
        agents = []
        for name in agent_names:
            details = cls.get_agent_details(world_name, name)
            if details:
                agents.append(details)
        return agents

File: backend/src/services/agent-filesystem-service.ts
/**
 * Character agents living inside one world (`worlds/{name}/agents/`), created at
 * run time by the Character Designer — distinct from the repo-level `agents/`
 * tree, though both share a layout and a parser. A character is never deleted,
 * only archived to `agents/_archived/{name}_{stamp}/`; the leading underscore is
 * what keeps it out of {@link listWorldAgents}.
 */

import { copyFileSync, existsSync, mkdirSync, readdirSync, renameSync, writeFileSync } from 'node:fs'
import { extname, join } from 'node:path'

import { getLogger } from '@/infrastructure/logging/logger'
import { parseAgentConfig, REQUIRED_AGENT_FILES } from '@/sdk/parsing/agent-config'
import { MtimeCache, WorldService } from './world-service'

const logger = getLogger('AgentFilesystemService')

const ARCHIVE_DIR = '_archived'

// `name` and `folderName` currently always hold the same string; consumers
// compare against both, and `name` is where a real display name would go.
export interface WorldAgentDetails {
  name: string
  folderName: string
  inANutshell: string
  /** Absolute path to the profile image, or `null` when there is none. */
  profilePic: string | null
}

// `%Y%m%d_%H%M%S` in UTC — the archive folder's collision suffix.
function utcStamp(when: Date): string {
  const iso = when.toISOString()
  return `${iso.slice(0, 4)}${iso.slice(5, 7)}${iso.slice(8, 10)}_${iso.slice(11, 13)}${iso.slice(14, 16)}${iso.slice(17, 19)}`
}

export class AgentFilesystemService {
  private readonly worlds: WorldService

  constructor(worldsDir: string, cache: MtimeCache = new MtimeCache()) {
    this.worlds = new WorldService(worldsDir, cache)
  }

  private agentsDir(worldName: string): string {
    return join(this.worlds.getWorldPath(worldName), 'agents')
  }

  private agentPath(worldName: string, agentName: string): string {
    return join(this.agentsDir(worldName), agentName)
  }

  // Both identity files are written even when empty: a folder holding neither is
  // not an agent, so the character would be in the room roster but no listing.
  createAgent(
    worldName: string,
    agentName: string,
    inANutshell: string,
    characteristics: string,
    profilePicPath?: string | null,
  ): void {
    const agentPath = this.agentPath(worldName, agentName)
    // `recursive` is forgiving of hand-made worlds with no `agents/` directory.
    mkdirSync(agentPath, { recursive: true })

    writeFileSync(join(agentPath, 'in_a_nutshell.md'), inANutshell, 'utf-8')
    writeFileSync(join(agentPath, 'characteristics.md'), characteristics, 'utf-8')

    if (profilePicPath && existsSync(profilePicPath)) {
      copyFileSync(profilePicPath, join(agentPath, `profile${extname(profilePicPath)}`))
    }

    logger.info(`Created agent '${agentName}' in world '${worldName}'`)
  }

  /**
   * Move a character into `agents/_archived/{name}_{stamp}/`; `false` when there
   * was no such character. Two archives of one name inside the same second
   * collide and `renameSync` throws `ENOTEMPTY`, which surfaces as a failed
   * archive — better than a character silently buried in another one's folder.
   */
  archiveAgent(worldName: string, agentName: string): boolean {
    const agentPath = this.agentPath(worldName, agentName)
    if (!existsSync(agentPath)) {
      logger.warning(`Agent '${agentName}' not found in world '${worldName}'`)
      return false
    }

    const archivePath = join(this.agentsDir(worldName), ARCHIVE_DIR)
    mkdirSync(archivePath, { recursive: true })

    const archivedAgentPath = join(archivePath, `${agentName}_${utcStamp(new Date())}`)
    renameSync(agentPath, archivedAgentPath)

    logger.info(`Archived agent '${agentName}' in world '${worldName}' to ${archivedAgentPath}`)
    return true
  }

  /**
   * Folder names of the world's live characters: at least one identity file, no
   * leading `_`. Sorted deliberately — raw directory order is filesystem-
   * dependent, so the model would see a different roster per machine.
   */
  listWorldAgents(worldName: string): string[] {
    const agentsDir = this.agentsDir(worldName)
    if (!existsSync(agentsDir)) return []

    return readdirSync(agentsDir, { withFileTypes: true })
      .filter((entry) => entry.isDirectory() && !entry.name.startsWith('_'))
      .map((entry) => entry.name)
      .filter((name) =>
        REQUIRED_AGENT_FILES.some((file) => existsSync(join(agentsDir, name, file))),
      )
      .sort()
  }

  // Delegated to {@link parseAgentConfig} so there is one implementation of
  // "what an agent folder contains", at the cost of a few discarded reads.
  getAgentDetails(worldName: string, agentName: string): WorldAgentDetails | null {
    const agentPath = this.agentPath(worldName, agentName)
    const config = parseAgentConfig(agentPath)
    if (config === null) return null

    return {
      name: agentName,
      folderName: agentName,
      inANutshell: config.inANutshell ?? '',
      // An absolute path, not a filename: the frontend serves the image from it.
      profilePic: config.profilePic === null ? null : join(agentPath, config.profilePic),
    }
  }

  /** {@link listWorldAgents} resolved to details; vanished folders are dropped. */
  listWorldAgentsWithDetails(worldName: string): WorldAgentDetails[] {
    const details: WorldAgentDetails[] = []
    for (const name of this.listWorldAgents(worldName)) {
      const detail = this.getAgentDetails(worldName, name)
      if (detail !== null) details.push(detail)
    }
    return details
  }
}
